File: WinDiskWriter GUI/Classes/CommandLine/CommandLine.m
//
//  CommandLine.m
//  windiskwriter
//
//  Created by Macintosh on 26.01.2023.
//  Copyright © 2023 TechUnRestricted. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "CommandLine.h"

@implementation CommandLine

+ (CommandLineData *_Nullable)execute: (NSString *)executable
                            arguments: (NSArray *_Nullable)arguments {
        
    @try {
        NSTask *task = [[NSTask alloc] init];
        
        NSPipe *standartdPipe = [NSPipe pipe];
        [task setStandardOutput: standartdPipe];
        
        NSPipe *errorPipe = [NSPipe pipe];
        [task setStandardError: errorPipe];
        
        [task setLaunchPath: executable];
        
        if (arguments) {
            [task setArguments: arguments];
        }
        
        NSFileHandle *fileHandleStandardPipe = [standartdPipe fileHandleForReading];
        NSFileHandle *fileHandleErrorPipe = [errorPipe fileHandleForReading];
        
        [task launch];
        [task waitUntilExit];
        
        CommandLineData *commandLineData = [[CommandLineData alloc] initWithProcessIdentifier: [task processIdentifier]
                                                                            terminationStatus: [task terminationStatus]
                                                                            terminationReason: [task terminationReason]
                                                                                 standardData: [fileHandleStandardPipe readDataToEndOfFile]
                                                                                    errorData: [fileHandleErrorPipe readDataToEndOfFile]];
        
        return commandLineData;
    } @catch (NSException *exception) {
        /* An error occurred while executing a terminal command */
    }
    
    return NULL;
}

@end

File: WinDiskWriter/Services/CommandLine/CommandLine.swift
//
//  CommandLine.swift
//  WinDiskWriter
//
//  Created by Macintosh on 28.04.2024.
//

import Foundation

class CommandLine {
    private init() { }

    @discardableResult
    static func execute(executable: String, arguments: [String]? = nil) throws -> CommandLineResult {
        let task = Process()

        let standardPipe = Pipe()
        task.standardOutput = standardPipe

        let errorPipe = Pipe()
        task.standardError = errorPipe

        task.launchPath = executable
        task.arguments = arguments

        if let exception = LegacyExceptionHandler.catchException({
            task.launch()
            task.waitUntilExit()
        }) {
            throw CommandLineError.objectiveCException(errorString: exception.reason)
        }

        let standardOutputData = standardPipe.fileHandleForReading.readDataToEndOfFile()
        let standardErrorData = errorPipe.fileHandleForReading.readDataToEndOfFile()

        let commandLineResult = CommandLineResult(
            standardData: standardOutputData,
            errorData: standardErrorData,
            terminationStatus: task.terminationStatus
        )

        return commandLineResult
    }
}